File: npm/src/rules/shared/localAntecedent.js
// @ts-check
import { overlap, topicsOf } from "../../fingerprint/topics.js";

/** @param {import("../../fingerprint/build.js").SentencePrint} sentence */
export function hasLocalAntecedent(sentence) {
  const marker = sentence.deixis[0];
  const start = sentence.text.indexOf(marker);
  const before = start > 0 ? topicsOf(sentence.text.slice(0, start)) : new Set();
  if (marker.startsWith("이것")) return before.size > 0;
  if (marker.startsWith("해당 ") || marker.startsWith("이러한 ")) return overlap(before, topicsOf(marker)) > 0;
  return false;
}

File: src/hanlint/rules/shared/localAntecedent.py
"""지시어 규칙 둘이 함께 쓰는 같은 문장 안의 선행어 판정."""

from __future__ import annotations

from ...fingerprint import SentencePrint
from ...fingerprint.topics import overlap, topicsOf


def hasLocalAntecedent(sentence: SentencePrint) -> bool:
    """지시어보다 앞선 같은 문장 절에 이름이 있으면 되돌아갈 필요가 없다."""
    marker = sentence.deixis[0]
    start = sentence.text.find(marker)
    before = topicsOf(sentence.text[:start]) if start > 0 else frozenset()
    if marker.startswith("이것"):
        return bool(before)
    if marker.startswith("해당 "):
        return overlap(before, topicsOf(marker)) > 0.0
    return False
